Sources/PackageGraph/Pubgrub/PubgrubDependencyResolver.swift

```diff
@@ -1216,41 +1216,47 @@ internal final class PubGrubPackageContainer {
+                    terms.append(Term(not: dependencyNode, constraintRequirement))
                     return try Incompatibility(terms, root: root, cause: .dependency(node: node))
                 }
-            }.joined())
+            }
         }
 
-        let (lowerBounds, upperBounds) = try self.computeBounds(for: node,
-                                                                constraints: constraints,
-                                                                startingWith: version)
+        let (lowerBounds, upperBounds) = try self.computeBounds(
+            for: node,
+            constraints: constraints,
+            startingWith: version
+        )
 
-        return try constraints.flatMap { constraint in
+        return try constraints.flatMap { constraint -> [Incompatibility] in
+            // We only have version-based requirements at this point.
+            guard case .versionSet(let constraintRequirement) = constraint.requirement else {
+                throw InternalError("Unexpected unversioned requirement: \(constraint)")
+            }
             return try constraint.nodes().map { constraintNode in
                 var terms: OrderedSet<Term> = []
-                let lowerBound = lowerBounds[constraintNode] ?? "0.0.0"
-                let upperBound = upperBounds[constraintNode] ?? Version(version.major + 1, 0, 0)
-                assert(lowerBound < upperBound)
 
-                // We only have version-based requirements at this point.
-                guard case .versionSet(let vs) = constraint.requirement else {
-                    throw InternalError("Unexpected unversioned requirement: \(constraint)")
+                // Make a record for this dependency so we don't have to recompute the bounds when the selected version falls within the bounds.
+                if let lowerBound = lowerBounds[constraintNode], let upperBound = upperBounds[constraintNode] {
+                    assert(lowerBound < upperBound)
+                    let requirement: VersionSetSpecifier = .range(lowerBound ..< upperBound)
+                    self.emittedIncompatibilities[constraintNode] = requirement.union(emittedIncompatibilities[constraintNode] ?? .empty)
                 }
 
-                let requirement: VersionSetSpecifier = .range(lowerBound ..< upperBound)
-                terms.append(Term(node, requirement))
-                terms.append(Term(not: constraintNode, vs))
-
-                // Make a record for this dependency so we don't have to recompute the bounds when the selected version falls within the bounds.
-                self.emittedIncompatibilities[constraintNode] = requirement.union(emittedIncompatibilities[constraintNode] ?? .empty)
+                // the package version requirement
+                terms.append(Term(node, .exact(version)))
+                // the dependency's version requirement
+                terms.append(Term(not: constraintNode, constraintRequirement))
```

Comment:
this is the heart of the change, there a couple of things going on here that seem wrong:

1: `requirement` is used for two different purposes: to record the dependency version (via ` self.emittedIncompatibilities`) but also as the resolved version for `node` (via `terms.append(Term(node, requirement))`), which iiuc does not really make sense as it mixed the dependency (`constraintNode`) version with the package (`node`) version.

2.  `requirement` is looked up on the bounds computed for the dependencies using `constraintNode` but if not found, defaults to 0.0.0 < version(major+1, 0, 0). the defaulting is wrong in two ways: it does not work for pre-release versions since it strips the rep-release extensions and it is based on `version` which is the version of `node` not `constraintNode`

the changes is to always set the `node` term to the exact version requested which works for pre-release and non-pre-release. it does however break a bunch of tests that recorded the 0.0.0 <  version(major+1, 0, 0) behavior, but maybe that is okay since they are testing for wrong behavior?

@SDGGiesbrecht very interested in your 👀 on this, as you are probably more familiar with this code / logic than I am. I actually hope I am wrong and this is not as wrong as I think it is :D

Reply:
I don't have much knowledge here, but I would not be surprised if some of the tests are conceptually wrong and just verifying the implemented behavior.  So I wouldn't be concerned about having to change the tests.

Reply:
Taking a closer look now. I will get back to you.

Reply:
I believe both your observations are correct, @tomerd, and same goes for your solutions.

Whatever situation used to hit the defaults will now result in nothing recorded into `emittedIncompatibilities`. At first that concerned me, but after tracking where that leads, it feels safe, even if a little odd. The only two times `emittedIncompatibilities` is read from are:
- To skip processing a node that has already been handled. Such nodes will be re‐processed now, but I don’t think that would be a problem.
- To not trust a pin if we have already encountered anything odd. This too only prevents a shortcut, as the whole graph will be traversed either way, just on the one hand not necessarily trying the most probable solution first when obvious, or on the other hand not preloading alternatives when overestimating the probability that a solution will pan out.

I would be curious to know _how_ the defaults were reachable in the first place, and if it could be done from the wild or only by mocked tests that bypass earlier checks.

> I actually hope I am wrong and this is not as problematic as I think it is :D

Yes, that code appears to have been embarrassingly invalid. I think it snuck by everyone for so long because without the proper incompatibility here, it would just fail to record a short‐circuit opportunity and merely be slower, and because the nonsensical extra incompatibility would usually end up so far afield that it would be completely disjoint from the actual problem space.